### DISCUSSION TITLE
Modern Keycloak JS Adapter

### DISCUSSION THREAD

**Discussion by DaSchTour** [keycloak/keycloak#8623] [Ideas]
I came across this issue while working on a mobile app using Capacitor. The current implementation of keycloak-js is hard to extend and the code is probably also hard to maintain. For Capacitor there is an alternative implementation of keycloak-js: https://github.com/JohannesBauer97/keycloak-ionic because a lot of functions would need to be copied to. From this experience I started this and ported keycloak-js to typescript https://github.com/footageone/keycloak-js/tree/main/src
It's now modular and could also be release as an ESM version. It also includes the capacitor adapters. But actually I would suggest to create separate packages for every adapter, as each adapter has different requirements.
There is still a lot of work to improve the keycloak-js code make it up to date.

**Comment by stianst** [comment]
@jonkoops FIY

**Comment by stianst** [comment]
I'm all for revamping the JS adapter, both in terms of how it is implemented/built, but also would like to see some improvements on its internals/APIs as well.

Would probably make sense to start it as a completely new project (separate git repo), which will make it possible to break the API and change how it behaves without worrying about backwards compatiblity.

**Reply by DaSchTour** [reply]
It's really time and I would like to start immediately. I'm fighting all week to get the Login working for my Ionic/Capacitor 3 app to work. It's hard to debug and understand why the redirects sometimes fail, sometimes parsing of the response fails and I have to edit and publish the adapter to have it work with keycloak-angular module. It's a real pain and when looking on the ecosystem I have the impression that keycloak isn't used that often in such scenarios, probably because it's nearly impossible to extend and adapt keycloak-js to different scenarios.

**Reply by jonkoops** [reply]
> Would probably make sense to start it as a completely new project (separate git repo), which will make it possible to break the API and change how it behaves without worrying about backwards compatiblity.

Yes, this is also something I think we are going to have to go for. The current API is hard to work with and will not really a good fit for a lot of use cases. Besides that the internals are an absolute mess and should need heavy refactoring.

At this point I am too bogged down with the new admin console to realistically contribute on a new client, but this is definitely something on my list long term.

I also want to consolidate various community efforts into this project, including but not limited to:

- An official OpenID connect JavaScript client to replace `keycloak-js` written in TypeScript which works in the browser,  React Native, NodeJS and Deno
- Lightweight companion libraries that follow the best practices for frameworks such as Angular, React and VueJS

**Reply by stianst** [reply]
First step here is really to start on a more detailed proposal. If someone from the community wants to take that on that'd be great, otherwise this will end up in the backlog for a while.

It's also quite relevant to reassess the security around a SPA integration. There's different approaches mentioned in oauth best practices, and we only support the least secure approach today. There's also some techniques that can be used to make it more secure with the current approach by sandboxing the tokens more (if I don't remember incorrectly it's called service workers).

Final point to make here is that we should also review if there are better oauth/oidc libraries out there now that can either be used as is, or be used within a thinner wrapper library from our end.

**Reply by jonkoops** [reply]
Yes, we certainly need to work on a proposal taking more modern standards into account. I am absolutely in favor of looking around to see if there are existing projects out there that can provide value in this regard (suggestions welcome).

**Reply by DaSchTour** [reply]
Maybe it would also help a lot to give some guidelines. I wanted to use a generic OAuth library with Capacitor but got somehow lost in terminology and in understanding what do I need to provide for the configuration. I often see me confronted with the problem, that example configurations for Google, Azure, Auth0 but rarely for Keycloak.

Maybe it's a good idea to check how existing solutions can be used.

For example
- For capacitor: https://github.com/moberwasserlechner/capacitor-oauth2
- For angular: https://github.com/manfredsteyer/angular-oauth2-oidc

**Reply by jonkoops** [reply]
I am a big fan of the work of Manfred on `angular-oauth2-oidc`, unfortunately it is largely Angular specific.

**Reply by DaSchTour** [reply]
Well, from my experience of working with keycloak-ionic there are two core parts in working with keycloak-js is generating the URLs and parsing the response. As one of the few still maintained generic solutions I found https://github.com/salte-auth/salte-auth
But as I see the handlers expect a full response with the tokens.

For me currently the hardest part is to understand how parsing of the data and retrieving consecutive requests in the callbacks work and how SSO checks and the silent renewal of tokens should work. Probably having this parts split into small modules would make it a lot easier to create implementations for different libraries.
